src/tests/ftest/util/spdk.py

```diff
@@ -0,0 +1,120 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+  The Government's rights to use, modify, reproduce, release, perform, display,
+  or disclose this software are subject to the terms of the Apache License as
+  provided in Contract No. B609815.
+  Any reproduction of computer software, computer software documentation, or
+  portions thereof marked with this legend must also reproduce the markings.
+'''
+from __future__ import print_function
+
+import getpass
+import threading
+import Queue
+
+from ssh_connection import Ssh
+
+SPDK_SETUP_SCRIPT = "/opt/daos/spdk/scripts/setup.sh"
+
+class SpdkFailed(Exception):
+    """ SPDK Setup/Cleanup did not work """
+
+def nvme_setup_thread(result_queue, hostname, debug=True):
+    """
+    nvme setup thread function
+    Arge:
+        result_queue: Queue for PASS/FAIL
+        hostname: server name
+        debug: print the debug message, default True
+    """
+    host = Ssh(hostname, debug=debug)
+    host.connect()
+    cmd = ["sudo /usr/bin/mkdir -p /opt/daos",
+           "sudo /usr/bin/git clone https://github.com/spdk/spdk.git" \
+           " /opt/daos/spdk",
+           "sudo HUGEMEM=4096 TARGET_USER=\"{0}\" {1}"
+           .format(getpass.getuser(), SPDK_SETUP_SCRIPT),
+           "sudo /usr/bin/chmod 777 /dev/hugepages",
+           "sudo /usr/bin/chmod 666 /dev/uio*",
+           "sudo /usr/bin/chmod 666 \
+           /sys/class/uio/uio*/device/config",
+           "sudo /usr/bin/chmod 666 \
+           /sys/class/uio/uio*/device/resource*",
+           "sudo /usr/bin/rm -f /dev/hugepages/*"]
+    rccode = host.call(" && ".join(cmd))
+    if rccode is not 0:
+        print("--- FAIL --- Failed SPDK setup command on {} with Error {}"
+              .format(hostname, rccode))
+        result_queue.put("FAIL")
+    result_queue.put("PASS")
+    host.disconnect()
+
+def nvme_cleanup_thread(result_queue, hostname, debug=True):
+    """
+    nvme cleanup thread function
+    Args:
+        result_queue: Queue for PASS/FAIL
+        hostname: server name
+        debug: print the debug message, default True
+    """
+    host = Ssh(hostname, debug=debug)
+    host.connect()
+    cmd = ["sudo HUGEMEM=4096 TARGET_USER=\"{0}\" {1} reset"
+           .format(getpass.getuser(), SPDK_SETUP_SCRIPT),\
+           "sudo /usr/bin/rm -f /dev/hugepages/*",
+           "sudo /usr/bin/rm -rf /opt/daos/"]
+    rccode = host.call(" && ".join(cmd))
+    if rccode is not 0:
+        print("--- FAIL --- Failed SPDK setup command on {} with Error {}"
+              .format(hostname, rccode))
+        result_queue.put("FAIL")
+    result_queue.put("PASS")
+    host.disconnect()
+
+def nvme(hostlist, operation="cleanup"):
```

Comment:
Could we rename this function to something more descriptive?  At first I found it a little odd that the same method was being used in setUp() and tearDown().  Given the binary nature of the operation, should the second argument be a bool, e.g. setup=True, and handle the string setting internally, e.g. operation = "setup" if setup else "cleanup"?

Reply:
I would even advocate simple wrapper functions, `nvme_setup()` and `nvme_teardown()` (as just an example) that handled the calling of the generic function with the correct args/flags to indicate setup/teardown, just to make what is happening in the caller more clear.